Lake/Build/Module.lean

```diff
@@ -136,79 +121,46 @@ def Module.recBuildLeanCore (mod : Module)
-`.c` file if the module is not lean-only or the requested artifact is `c`.
-Returns a build job producing the requested artifact.
--/
-def Module.recBuildLean (mod : Module) (art : LeanArtifact)
-: IndexBuildM (BuildJob (if art = .none then Unit else FilePath)) := do
-  let leanOnly := mod.isLeanOnly ∧ art ≠ .c
-  let modJob ← mod.recBuildLeanCore leanOnly
-  -- Save All Resulting Jobs & Return Requested One
-  store mod.leanBin.key modJob
-  let oleanJob ← modJob.bindSync fun _ depTrace =>
-    return (mod.oleanFile, mixTrace (← computeTrace mod.oleanFile) depTrace)
-  store mod.olean.key <| oleanJob
-  let ileanJob ← modJob.bindSync fun _ depTrace =>
-    return (mod.ileanFile, mixTrace (← computeTrace mod.ileanFile) depTrace)
-  store mod.ilean.key <| ileanJob
-  if h : leanOnly then
-    have : art ≠ .c := h.2
-    return match art with
-    | .none => modJob
-    | .olean => oleanJob
-    | .ilean => ileanJob
-  else
-    let cJob ← modJob.bindSync fun _  _ =>
-      return (mod.cFile, mixTrace (← computeTrace mod.cFile) (← getLeanTrace))
-    store mod.c.key cJob
-    return match art with
-    | .none => modJob
-    | .olean => oleanJob
-    | .ilean => ileanJob
-    | .c => cJob
+  extraDepJob.bindAsync fun _ _ => do
+  importJob.bindAsync fun _ importTrace => do
+  modDynlibsJob.bindAsync fun modDynlibs modTrace => do
+  externDynlibsJob.bindSync fun externDynlibs externTrace => do
+    let depTrace := importTrace.mix <| modTrace.mix externTrace
+    /-
+    Requirements:
+    * Lean wants the external library symbols before module symbols.
+    * Unix requires the file extension of the dynlib.
+    * For some reason, building from the Lean server requires full paths.
+      Everything else loads fine with just the augmented library path.
+    * Linux still needs the augmented path to resolve nested dependencies in dynlibs.
+    -/
+    let dynlibPath := libDirs ++ externDynlibs.filterMap (·.dir?) |>.toList
+    let dynlibs := externDynlibs.map (·.path) ++ modDynlibs.map (·.path)
+    let trace ← mod.buildUnlessUpToDate dynlibPath dynlibs depTrace
+    return ((), trace)
 
 /-- The `ModuleFacetConfig` for the builtin `leanBinFacet`. -/
 def Module.leanBinFacetConfig : ModuleFacetConfig leanBinFacet :=
-  mkFacetJobConfig (·.recBuildLean .none)
+  mkFacetJobConfig (·.recBuildLeanCore)
 
 /-- The `ModuleFacetConfig` for the builtin `oleanFacet`. -/
 def Module.oleanFacetConfig : ModuleFacetConfig oleanFacet :=
-  mkFacetJobConfig (·.recBuildLean .olean)
+  mkFacetJobConfigSmall fun mod => do
+    (← mod.leanBin.fetch).bindSync fun _ depTrace =>
+      return (mod.oleanFile, mixTrace (← computeTrace mod.oleanFile) depTrace)
 
 /-- The `ModuleFacetConfig` for the builtin `ileanFacet`. -/
 def Module.ileanFacetConfig : ModuleFacetConfig ileanFacet :=
-  mkFacetJobConfig (·.recBuildLean .ilean)
+  mkFacetJobConfigSmall fun mod => do
+    (← mod.leanBin.fetch).bindSync fun _ depTrace =>
+      return (mod.ileanFile, depTrace)
```

Comment:
```suggestion
      return (mod.ileanFile, mixTrace (← computeTrace mod.ileanFile) depTrace)
```

The trace should include the ilean trace like it originally did. Should it not?

Reply:
I don't think so.  If we use only `computeTrace` on its own, then we get a content-based hash where reverse dependencies are recompiled when the `.ilean` file changes.  (Or not, if e.g. only a typo in a comment was fixed, which doesn't change affect the `.ilean` file.)  Or we use `depTrace` directly, then we're assuming that the `.ilean` file is determined by the `.lean` file, its dependencies, and the Lean version.  Both options sound correct to me.

On second thought, we should probably apply the same logic to the olean files then as well.  We could avoid hashing *gigabytes* of olean files.  (Only downside is that you'll need `lake clean` if you write an impure Lean file, see #86.)

Reply:
>  Or we use `depTrace` directly, then we're assuming that the `.ilean` file is determined by the `.lean` file, its dependencies, and the Lean version. Both options sound correct to me.

That is not what `depTrace` is in this context. `depTrace` is just the module dependencies (import oleans and shared libraries) plus the module trace (its olean and ilean) - a detail I had previously forgotten.

As to question of what should be the trace. I think sticking with the binary file plus the dependency trace is right. Like you mentioned, this means that some small changes will not cause a rebuild (e.g., space changes, a typo) and impure changes will, which I think is ideal.

Reply:
@Kha Note that hashing the olean is not just about preventing rebuilds (which, admittedly, is rare). It also serves to force rebuilds in case were impure changes are stored within the olean (e.g., results of `include_str`, impure environment extensions, etc.). Thus, such checks need to be made even if the olean input hashes do not differ.

> A quick perf run suggests that we spend 70% of a no-op mathlib4 build hashing .olean

I would expect that? After all, hashing is pretty much the only thing a no-op build really does (after loading the Lake workspace) and we usually only hash two files without native compilation (the lean and olean -- and the olean is generally much bigger). The bigger question is: how long is such a no-op build (and how does it compare to a normal build)?